cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -0,0 +1,268 @@
+ * @param gather_map_size Size of the gather map.
+ * @param nullify_out_of_bounds Nullify values in `gather_map` that are out of bounds
+ * @param stream CUDA stream on which to execute kernels
+ * @param mr Memory resource to use for all allocations
+ * @param prev_base_offsets (optional) The buffer backing the base offsets used in the gather map.
+ *
+ * @returns The gather_data struct needed to construct the gather map for the
+ *          next level of recursion.
+ *
+ */
+template <bool NullifyOutOfBounds, typename MapItType>
+gather_data make_gather_data(
+  cudf::lists_column_view const& source_column,
+  MapItType gather_map,
+  size_type gather_map_size,
+  cudaStream_t stream,
+  rmm::mr::device_memory_resource* mr,
+  rmm::device_uvector<size_type> prev_base_offsets = rmm::device_uvector<size_type>{})
+{
+  // size of the gather map is the # of output rows
+  size_type output_count = gather_map_size;
+  size_type offset_count = output_count + 1;
+
+  // offsets of the source column
+  size_type const* src_offsets{source_column.offsets().data<size_type>()};
+  int src_size = source_column.size();
+
+  // outgoing offsets.  these will persist as output from the entire gather operation
+  auto dst_offsets_c = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+  mutable_column_view dst_offsets_v = dst_offsets_c->mutable_view();
+  size_type* dst_offsets            = dst_offsets_v.data<size_type>();
+
+  // generate the compacted outgoing offsets.
+  auto count_iter = thrust::make_counting_iterator<size_type>(0);
+  thrust::plus<size_type> sum;
+  thrust::transform_exclusive_scan(
+    rmm::exec_policy(stream)->on(stream),
+    count_iter,
+    count_iter + offset_count,
+    dst_offsets,
+    [gather_map, output_count, src_offsets, src_size] __device__(size_type index) -> size_type {
+      // last offset index is always the previous offset_index + 1, since each entry in the gather
+      // map represents a virtual pair of offsets
+      size_type offset_index = index < output_count ? gather_map[index] : gather_map[index - 1] + 1;
+
+      // if this is an invalid index, this will be a NULL list
+      if (NullifyOutOfBounds && ((offset_index < 0) || (offset_index >= src_size))) { return 0; }
+
+      // the length of this list
+      return src_offsets[offset_index + 1] - src_offsets[offset_index];
+    },
+    0,
+    sum);
+
+  // For each span of offsets (each output row) we need to know the original offset value to build
+  // the gather map for the next level.  This data is temporary and will only be needed until the
+  // next level of recursion is done using it.
+  //
+  // additional optimization. if the "prev_base_offsets" buffer we have been passed is large enough
+  // to be used again, re-use it instead of allocating new memory. if it isn't big enough, release
+  // it.
+  if (prev_base_offsets.capacity() > 0) {
```

Comment:
This is very cool Dave.

Reply:
In practice it doesn't hit very often because it's rare that the # of offsets goes -down- as you go deeper in the hierarchy, but it can happen.